File: Wirtualny_Swiat_Java/Projekt_PO_Java/src/model/Organisms/Fox.java
package model.Organisms;

import model.Animal;
import model.Organism;
import model.Point;
import model.World;

import javax.swing.text.Position;
import java.awt.*;

public class Fox extends Animal {
    private Color foxColor = Color.ORANGE;
    private final int attempsToMove = 10;

    public Fox(World world, Point position){
        super(world, position, 3, 7);
    }

    @Override
    public Color getColor(){
        return foxColor;
    }

    @Override
    public String getNazwa(){
        return "Lis";
    }

    @Override
    public void Action(){
        increaseAge();
        Point currentPos = getPosition();

        for (int i = 0; i < attempsToMove; i++){
            int direction = random.nextInt(4);
            Point newPosition = new Point(currentPos.getX(), currentPos.getY());

            switch (direction){
                case 0 -> newPosition.setY(currentPos.getY() - 1);
                case 1 -> newPosition.setX(currentPos.getX() + 1);
                case 2 -> newPosition.setY(currentPos.getY() + 1);
                case 3 -> newPosition.setX(currentPos.getX() - 1);
            }

            if (shouldMove(newPosition)){
                world.tryToMoveOrganism(this, newPosition);
                return;
            }
        }
    }

    private boolean shouldMove(Point position){
        int x = position.getX();
        int y = position.getY();
        for (Organism o : world.getOrganisms()){
            if (x == o.getPosition().getX() && y == o.getPosition().getY() && o.getPower() > this.getPower()){
//                world.addEvent("Lis ucieka przed " + o.getNazwa() + " na pozycji(" + x + ", " + y + ")");
                return false;
            }
        }
        return true;
    }

    @Override
    public Organism makeNewOrganism(Point position){
        return new Fox(world, position);
    }
}

File: WirtualnySwiatPython/Animals/Fox.py
import random

from animal import Animal
from config import ORANGE
from point import Point


class Fox(Animal):
    def __init__(self, world, position):
        super().__init__(world, position, power=3, initiative=7)
        self.color = ORANGE
        self.attempsToMove = 10

    def getColor(self):
        return self.color

    def getNazwa(self):
        return "Lis"

    def makeNewOrganism(self, position):
        return Fox(self.world, position)

    def Action(self):
        self.increaseAge()
        position = self.getPosition()

        for i in range(self.attempsToMove):
            self.increaseAge()
            direction = random.randint(0, 3)

            new_x = position.GetX()
            new_y = position.GetY()

            if direction == 0:
                new_y -= 1
            elif direction == 1:
                new_x += 1
            elif direction == 2:
                new_y += 1
            elif direction == 3:
                new_x -= 1

            newPosition = Point(new_x, new_y)
            if self.shouldMove(newPosition):
                self.world.tryToMoveOrganism(self, newPosition)
                break

    def shouldMove(self, position):
        x = position.GetX()
        y = position.GetY()
        for organism in self.world.organisms:
            if x == organism.getPosition().GetX() and y == organism.getPosition().GetY() and organism.getPower() > self.getPower():
                return False

        return True